stl/src/filesystem.cpp

```diff
@@ -126,6 +126,40 @@ namespace {
         return __std_win_error{GetLastError()};
     }
 
+    [[nodiscard]] __std_win_error __stdcall _Get_file_id_by_handle(
+        const HANDLE _Handle, _Out_ FILE_ID_INFO* const _Id) noexcept {
+        __std_win_error _Last_error;
+        if (GetFileInformationByHandleEx(_Handle, FileIdInfo, _Id, sizeof(*_Id)) != 0) {
+            // if we could get FILE_ID_INFO, use that as the source of truth
+            return __std_win_error::_Success;
+        }
+
+        _Last_error = __std_win_error{GetLastError()};
+        switch (_Last_error) {
+        case __std_win_error::_Not_supported:
+        case __std_win_error::_Invalid_parameter:
+            break; // try more things
+        default:
+            return _Last_error; // real error, bail to the caller
+        }
+
+#ifndef _CRT_APP
+        // try GetFileInformationByHandle as a fallback
+        BY_HANDLE_FILE_INFORMATION _Info;
+        if (GetFileInformationByHandle(_Handle, &_Info) != 0) {
+            _Id->VolumeSerialNumber = _Info.dwVolumeSerialNumber;
+            _CSTD memcpy(&_Id->FileId.Identifier[0], &_Info.nFileIndexHigh, 8);
+            _CSTD memset(&_Id->FileId.Identifier[8], 0, 8);
+            return __std_win_error::_Success;
+        }
+
+        _Last_error = __std_win_error{GetLastError()};
+#endif // _CRT_APP
+
+        return _Last_error;
+    }
+
+
     struct _File_disposition_info_ex {

```

Comment:
There's a merge conflict now that #2714 has landed. It should be simple to resolve - you're adding code immediately above where I deleted `_File_disposition_info_ex`.